### PR TITLE
Feature refactor types gltf

### PR DESCRIPTION
**Pull request by saveman71**

### CODE REVIEW

File: include/lug/Graphics/Vulkan/Render/Pipeline.hpp [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+
+} // Render
+} // Vulkan
+} // Graphics
+} // lug
+
+// Make Pipeline Id hashable like a uint32_t
+namespace std {
+    template <> struct hash<lug::Graphics::Vulkan::Render::Pipeline::Id>
+    {
```
Comment (saveman71) [inline include/lug/Graphics/Vulkan/Render/Pipeline.hpp]: Brackets location (same below)

File: src/lug/Graphics/Render/Material.cpp [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+Material::Material(const std::string& name): Resource(Resource::Type::Material), _name(name) {
+    (void)_baseColorTexture;
+    (void)_metallicRoughnessTexture;
+    (void)_normalTexture;
+    (void)_occlusionTexture;
+    (void)_emissiveTexture;
+    (void)_constants;
+}
+
+
```
Comment (saveman71) [inline src/lug/Graphics/Render/Material.cpp]: Empty line

File: src/lug/Graphics/Vulkan/API/PipelineLayout.cpp [hunk 3]
```diff
@@ -87,8 +87,8 @@ std::unique_ptr<PipelineLayout> PipelineLayout::create(const Device* device) {
 
     if (!descriptorSetLayouts[0] || !descriptorSetLayouts[1]) {
         LUG_LOG.error("RendererVulkan: Can't create pipeline descriptor sets layout");
-        return nullptr;
-    }
+        return nullptr
```
Comment (saveman71) [inline src/lug/Graphics/Vulkan/API/PipelineLayout.cpp]: ???

File: src/lug/Graphics/Vulkan/Builder/Mesh.cpp [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+    int& meshMemoryTypeIndex,
+    VkDeviceSize& meshMemorySize,
+    Render::Mesh::PrimitiveSet::Attribute& attribute) {
+
+    auto& device = renderer->getDevice();
+
+    VkBufferUsageFlags usage;
+    if (attribute.type == lug::Graphics::Render::Mesh::PrimitiveSet::Attribute::Type::Indice) {
+        usage = VK_BUFFER_USAGE_INDEX_BUFFER_BIT;
+    }
```
Comment (saveman71) [inline src/lug/Graphics/Vulkan/Builder/Mesh.cpp]: } else

File: src/lug/Graphics/Vulkan/Builder/Mesh.cpp [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+    // Check the memory type of the buffer
+    // All the buffers should have the same memory type
+    // Because we create one device memory for all the buffers (So, with one type)
+    auto& requirements = buffer->getRequirements();
+    int attributeMemoryTypeIndex = API::DeviceMemory::findMemoryType(&device, requirements, VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT);
+    if (attributeMemoryTypeIndex != meshMemoryTypeIndex &&
+        meshMemoryTypeIndex != -1) {
+        LUG_LOG.error("Vulkan::Mesh::build: Buffers don't have the same memory type");
+        return nullptr;
+    }
```
Comment (saveman71) [inline src/lug/Graphics/Vulkan/Builder/Mesh.cpp]: } else

File: test/Graphics/Vulkan/Shaders.cpp [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+                    if (countTexCoord == 0) {
+                        materialPart.baseColorInfo = 0b11;
+                        materialPart.metallicRoughnessInfo = 0b11;
+                        materialPart.normalInfo = 0b11;
+                        materialPart.occlusionInfo = 0b11;
+                        materialPart.emissiveInfo = 0b11;
+
+                        pipelineIds.push_back(Vulkan::Render::Pipeline::Id::create(primitivePart, materialPart));
+                    } else {
```
Comment (saveman71) [inline test/Graphics/Vulkan/Shaders.cpp]: continue instead

File: include/lug/Graphics/Vulkan/Render/Material.hpp [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+     *
+     * @param[in]  name  The name of the Material
+     * @param[in]  type  The type of the Material
+     */
+    Material(const std::string& name);
+
+private:
+    API::DeviceMemory _deviceMemory;
+    API::Buffer _buffer;
+    //API::BufferView _bufferView;
```
Comment (saveman71) [inline include/lug/Graphics/Vulkan/Render/Material.hpp]: Do we still need this?

File: resources/shaders/forward/shader.frag [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+
+    //////////////////////////////////////////////////////////////////////
+    // CALCULATE THE FINAL METALLIC & ROUGHNESS
+    //////////////////////////////////////////////////////////////////////
+
+    float metallic = material.metallic;
+    float roughness = material.roughness;
+
+    #if TEXTURE_METALLIC_ROUGHNESS
+    vec3 textureValue = texture(texture_metallic_roughness, TEXTURE_METALLIC_ROUGHNESS_UV).rgb;
```
Comment (saveman71) [inline resources/shaders/forward/shader.frag]: `texture_value` if you want to be coherent?

File: resources/shaders/forward/shader.vert [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+// VERTEX SHADER
+//////////////////////////////////////////////////////////////////////////////
+
+#version 450
+
+//////////////////////////////////////////////////////////////////////////////
+// BLOCK OF STATIC INPUTS
+//////////////////////////////////////////////////////////////////////////////
+
+layout (push_constant) uniform modelBlock {
```
Comment (saveman71) [inline resources/shaders/forward/shader.vert]: snake_case

File: resources/shaders/forward/shader.vert [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+    mat4 transform;
+} model;
+
+layout(set = 0, binding = 0) uniform cameraBlock {
```
Comment (saveman71) [inline resources/shaders/forward/shader.vert]: snake_case

File: samples/base/CMakeLists.txt [hunk 11]
```diff
@@ -23,5 +23,5 @@ include(${PROJECT_SOURCE_DIR}/../Macros.cmake)
 
 lug_add_sample(base
                SOURCES main.cpp
-               DEPENDS system window
+               DEPENDS system window graphics core
```
Comment (saveman71) [inline samples/base/CMakeLists.txt]: Why?

File: src/lug/Graphics/Builder/Material.cpp [hunk 12]
```diff
@@ -0,0 +1,58 @@
+#include <lug/Graphics/Builder/Material.hpp>
+
+namespace lug {
+namespace Graphics {
+namespace Builder {
+
+Material::Material(Renderer& renderer) : _renderer(renderer) {}
+
+void Material::Material::setName(const std::string& name) {
```
Comment (saveman71) [inline src/lug/Graphics/Builder/Material.cpp]: Shouldn't the setters be in the inl file?

**Review by Quentin01 — APPROVED** [review]

**Pull request merged**